azurelinuxagent/common/utils/cryptutil.py

```diff
@@ -128,3 +132,12 @@ def bits_to_bytes(self, bits):
                 index = 7
         return bytes(byte_array)
 
+    def decrypt_secret(self, encrypted_password, private_key):
+        decoded = base64.b64decode(encrypted_password)
+        p = subprocess.Popen([self.openssl_cmd, 'cms', '-decrypt', '-inform', 'DER', '-inkey', private_key, '-in', '/dev/stdin'], stdout=subprocess.PIPE, stdin=subprocess.PIPE, stderr=subprocess.STDOUT)
+        p.stdin.write(decoded)
+        output = p.communicate()[0]
+        retcode = p.poll()
+        if retcode:
+            raise subprocess.CalledProcessError(retcode, "openssl cms -decrypt", output=output)
+        return output.decode('utf-16')
```

Comment:
Why not utf-8? Also, newline at the end.

Reply:
?  It's utf-16 and want the secret as-is... I don't want to add a newline to it.  Am I not understanding this comment?

---
In reply to: [194024673](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r194024673) [](ancestors = 194024673)

Reply:
I am just surprised if it is really utf-16, this would be the only time we use it.

Reply:
The newline is a pep-8 standard.

Reply:
Sorry, thought you wanted a newline at the end of the return, not the file which made no sense to me.  Fixed

---
In reply to: [194177678](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r194177678) [](ancestors = 194177678)

Reply:
Matt and I talked about this. The UTF-16 data here is incoming from FC via the host, so the agent has to convert _from_ UTF-16 to a native Python ustr. The surprising thing is that the far end of this transaction is _generating_ the payload and transmitting it in UTF-8. The transport channel from initiator to agent involves multiple layers of code which, for a variety of reasons, decrypt and re-encrypt this payload string at each layer transition. One layer of code is failing to treat the payload as an opaque sequence of octets; instead, it's treating it as a UTF-8 string, reading it into a .NET String (and re-encoding it to the native 16-bit WCHAR representation .NET uses), then grabbing the sequence of WCHARs and re-encrypting it for consumption by the next layer.

Matt's going to go on a Snark Hunt to find which layer is peeking inside the opaque bag of octets and get them to fix their bug. Until that happens, though, this `str.decode("utf-16")` has to remain.